File: ios/AirMoPub/MoPubBanner.m
/*
 * Copyright 2017 FreshPlanet
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#import "MoPubBanner.h"

@interface MoPubBanner () {
}

@property(nonatomic, assign) FREContext context;
@property(nonatomic, assign) BOOL autoRefreshEnabled;

@end

@implementation MoPubBanner

@synthesize context;

- (id)initWithContext:(FREContext)extensionContext adUnitId:(NSString*)adUnitId size:(CGSize)size {
    
    self = [super initWithAdUnitId:adUnitId size:size];
    
    if (self) {
        self.context = extensionContext;
        self.delegate = self;
        self.autoRefreshEnabled = YES;
    }
    return self;
}

- (void)dealloc {
    
    self.delegate = nil;
    [super dealloc];
}

- (float)getDisplayDensity {
    
    return [UIScreen mainScreen].scale;
}

+ (CGSize)getAdSizeFromSizeId:(int)sizeId {
    
    switch (sizeId) {
        case 1:
            return MOPUB_BANNER_SIZE;
        case 2:
            return MOPUB_MEDIUM_RECT_SIZE;
        case 3:
            return MOPUB_LEADERBOARD_SIZE;
        case 4:
            return MOPUB_WIDE_SKYSCRAPER_SIZE;
        default:
            return MOPUB_BANNER_SIZE;
    }
}

- (BOOL)getAutorefresh {
    
    return self.autoRefreshEnabled;
}

- (void)setAutorefresh:(BOOL)value {
    
    if (value && !self.autoRefreshEnabled)
        [self startAutomaticallyRefreshingContents];
    else if (!value && self.autoRefreshEnabled)
        [self stopAutomaticallyRefreshingContents];
    
    self.autoRefreshEnabled = value;
}

- (int)getPositionX {
    
    return (int)round(self.frame.origin.x * [self getDisplayDensity]);
}

- (int)getPositionY {
    
    return (int)round(self.frame.origin.y * [self getDisplayDensity]);
}

- (int)getFrameWidth {
    
    return (int)round(self.frame.size.width * [self getDisplayDensity]);
}

- (int)getFrameHeight {
    
    return (int)round(self.frame.size.height * [self getDisplayDensity]);
}

- (void)setPositionX:(int)value {
    
    CGRect frame = self.frame;
    frame.origin.x = (float)value / [self getDisplayDensity];
    self.frame = frame;
}

- (void)setPositionY:(int)value {
    
    CGRect frame = self.frame;
    frame.origin.y = (float)value / [self getDisplayDensity];
    self.frame = frame;
}

- (void)setFrameWidth:(int)value {
    
    CGRect frame = self.frame;
    frame.size.width = (float)value / [self getDisplayDensity];
    self.frame = frame;
}

- (void)setFrameHeight:(int)value {
    
    CGRect frame = self.frame;
    frame.size.height = (float)value / [self getDisplayDensity];
    self.frame = frame;
}

- (void)setAdSize:(int)value {
    
    CGSize size = [MoPubBanner getAdSizeFromSizeId:value];
    CGRect frame = self.frame;
    frame.size = size;
    self.frame = frame;
}

- (int)getCreativeWidth {
    
    return (int)round([self adContentViewSize].width * [self getDisplayDensity]);
}

- (int)getCreativeHeight {
    
    return (int)round([self adContentViewSize].height * [self getDisplayDensity]);
}

- (void)loadBanner {
    
    [self loadAd];
}

- (void)showBanner {
    
    [[UIApplication sharedApplication].keyWindow.rootViewController.view addSubview:self];
}

- (void)removeBanner {
    
    [self removeFromSuperview];
}

- (void)refresh {
    
    [self forceRefreshAd];
}

- (UIViewController*)viewControllerForPresentingModalView {
    
    return [UIApplication sharedApplication].keyWindow.rootViewController;
}

- (void)adViewDidLoadAd:(MPAdView*)view {
    
    FREDispatchStatusEventAsync(context, (const uint8_t*)"", (const uint8_t*)"bannerLoaded");
}

- (void)adViewDidFailToLoadAd:(MPAdView*)view {
    
    FREDispatchStatusEventAsync(context, (const uint8_t*)"", (const uint8_t*)"bannerFailedToLoad");
}

- (void)willPresentModalViewForAd:(MPAdView*)view {
    
    FREDispatchStatusEventAsync(context, (const uint8_t*)"", (const uint8_t*)"bannerAdClicked");
}

- (void)didDismissModalViewForAd:(MPAdView*)view {
    
    FREDispatchStatusEventAsync(context, (const uint8_t*)"", (const uint8_t*)"bannerAdClosed");
}

@end

# pragma mark - C/as interface

DEFINE_ANE_FUNCTION(banner_init) {
    
    NSString* adUnitId = FPANE_FREObjectToNSString(argv[0]);
    
    CGSize adType;
    int32_t sizeId;
    FREGetObjectAsInt32(argv[1], &sizeId);
    adType = [MoPubBanner getAdSizeFromSizeId:sizeId];
    
    MoPubBanner* banner = [[MoPubBanner alloc] initWithContext:context adUnitId:adUnitId size:adType];
    FRESetContextNativeData(context, (__bridge void*)(banner));
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_getAutorefresh) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_BOOLToFREObject([banner getAutorefresh]);
}

DEFINE_ANE_FUNCTION(banner_setAutorefresh) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setAutorefresh:FPANE_FREObjectToBool(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_getPositionX) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_IntToFREObject([banner getPositionX]);
}

DEFINE_ANE_FUNCTION(banner_getPositionY) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_IntToFREObject([banner getPositionY]);
}

DEFINE_ANE_FUNCTION(banner_getFrameWidth) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_IntToFREObject([banner getFrameWidth]);
}

DEFINE_ANE_FUNCTION(banner_getFrameHeight) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_IntToFREObject([banner getFrameHeight]);
}

DEFINE_ANE_FUNCTION(banner_setPositionX) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setPositionX:(int)FPANE_FREObjectToInt(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_setPositionY) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setPositionY:(int)FPANE_FREObjectToInt(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_setFrameWidth) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setFrameWidth:(int)FPANE_FREObjectToInt(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_setFrameHeight) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setFrameHeight:(int)FPANE_FREObjectToInt(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_setAdSize) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setAdSize:(int)FPANE_FREObjectToInt(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_getCreativeWidth) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_IntToFREObject([banner getCreativeWidth]);
}

DEFINE_ANE_FUNCTION(banner_getCreativeHeight) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_IntToFREObject([banner getCreativeHeight]);
}

DEFINE_ANE_FUNCTION(banner_loadBanner) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner loadBanner];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_showBanner) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner showBanner];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_removeBanner) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner removeBanner];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_refresh) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner refresh];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_setKeywords) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setKeywords:FPANE_FREObjectToNSString(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_setTesting) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    [banner setTesting:FPANE_FREObjectToBool(argv[0])];
    
    return NULL;
}

DEFINE_ANE_FUNCTION(banner_getTesting) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    return FPANE_BOOLToFREObject(banner.testing);
}

DEFINE_ANE_FUNCTION(banner_moveToDefaultPosition) {
    
    MoPubBanner* banner;
    FREGetContextNativeData(context, (void**)&banner);
    if (banner == nil)
        return FPANE_CreateError(@"Context does not have a Banner", 0);
    
    CGRect screenBounds = [[UIScreen mainScreen] bounds];
    CGFloat screenScale = [[UIScreen mainScreen] scale];
    CGSize screenSize = CGSizeMake(screenBounds.size.width * screenScale, screenBounds.size.height * screenScale);
    
    [banner setPositionX:(screenSize.width - [banner getFrameWidth]) / 2];
    [banner setPositionY:(screenSize.height - [banner getFrameHeight])];
    
    return NULL;
}

# pragma mark - list functions

void MoPubBannerListFunctions(const FRENamedFunction** functionsToSet, uint32_t* numFunctionsToSet) {
    
    static FRENamedFunction functions[] = {
            MAP_FUNCTION(banner_init, NULL),
            MAP_FUNCTION(banner_getAutorefresh, NULL),
            MAP_FUNCTION(banner_setAutorefresh, NULL),
            MAP_FUNCTION(banner_getPositionX, NULL),
            MAP_FUNCTION(banner_getPositionY, NULL),
            MAP_FUNCTION(banner_getFrameWidth, NULL),
            MAP_FUNCTION(banner_getFrameHeight, NULL),
            MAP_FUNCTION(banner_setPositionX, NULL),
            MAP_FUNCTION(banner_setPositionY, NULL),
            MAP_FUNCTION(banner_setFrameWidth, NULL),
            MAP_FUNCTION(banner_setFrameHeight, NULL),
            MAP_FUNCTION(banner_setAdSize, NULL),
            MAP_FUNCTION(banner_getCreativeWidth, NULL),
            MAP_FUNCTION(banner_getCreativeHeight, NULL),
            MAP_FUNCTION(banner_loadBanner, NULL),
            MAP_FUNCTION(banner_showBanner, NULL),
            MAP_FUNCTION(banner_removeBanner, NULL),
            MAP_FUNCTION(banner_refresh, NULL),
            MAP_FUNCTION(banner_setKeywords, NULL),
            MAP_FUNCTION(banner_setTesting, NULL),
            MAP_FUNCTION(banner_getTesting, NULL),
            MAP_FUNCTION(banner_moveToDefaultPosition, NULL)
    };
    
    *numFunctionsToSet = sizeof(functions) / sizeof(FRENamedFunction);
    *functionsToSet = functions;
}

File: android/lib/src/main/java/com/freshplanet/ane/AirMoPub/MoPubBanner.java
/*
 * Copyright 2017 FreshPlanet
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.freshplanet.ane.AirMoPub;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import com.adobe.fre.FREContext;
import com.adobe.fre.FREFunction;
import com.adobe.fre.FREObject;
import com.mopub.mobileads.MoPubErrorCode;
import com.mopub.mobileads.MoPubView;
import com.mopub.mobileads.MoPubView.BannerAdListener;

import java.util.HashMap;
import java.util.Map;

public class MoPubBanner extends MoPubView {
	
	public static final int BANNER_SIZE = 1;
	public static final int MEDIUM_RECT_SIZE = 2;
	public static final int LEADERBOARD_SIZE = 3;
	public static final int WIDE_SKYSCRAPER_SIZE = 4;
	
	private Context mContext;
	private FrameLayout.LayoutParams mLayoutParams;
	
	public MoPubBanner(Context context) {
		
		super(context);
		
		mContext = context;
		mLayoutParams = new FrameLayout.LayoutParams(0, 0, Gravity.TOP | Gravity.LEFT);
	}
	
	public MoPubBanner(Context context, AttributeSet attrs) {
		
		super(context, attrs);
	}
	
	public int getPosX() {
		
		return getUnscaledMetric(mLayoutParams.leftMargin);
	}
	
	public void setPosX(int x) {
		
		mLayoutParams.leftMargin = getScaledMetric(x);
		setLayoutParams(mLayoutParams);
	}
	
	public int getPosY() {
		
		return getUnscaledMetric(mLayoutParams.topMargin);
	}
	
	public void setPosY(int y) {
		
		mLayoutParams.topMargin = getScaledMetric(y);
		setLayoutParams(mLayoutParams);
	}
	
	public int getFrameWidth() {
		
		return getUnscaledMetric(mLayoutParams.width);
	}
	
	public void setFrameWidth(int width) {
		
		mLayoutParams.width = getScaledMetric(width);
		setLayoutParams(mLayoutParams);
	}
	
	public int getFrameHeight() {
		
		return getUnscaledMetric(mLayoutParams.height);
	}
	
	public void setFrameHeight(int height) {
		
		mLayoutParams.height = getScaledMetric(height);
		setLayoutParams(mLayoutParams);
	}
	
	public void setAdSize(int size) {
		
		int width, height;
		switch (size) {
			// from mopub-ios-sdk MPConstants.h
			case MEDIUM_RECT_SIZE:
				width = 300;
				height = 250;
				break;
			case LEADERBOARD_SIZE:
				width = 728;
				height = 90;
				break;
			case WIDE_SKYSCRAPER_SIZE:
				width = 160;
				height = 600;
				break;
			default: // BANNER_SIZE
				width = 320;
				height = 50;
				break;
		}
		
		setFrameWidth(width);
		setFrameHeight(height);
	}
	
	@Override
	public void destroy() {
		
		super.destroy();
		mContext = null;
		mLayoutParams = null;
	}
	
	public void moveToDefaultPosition() {
		
		setPosX(0);
		setPosY(0);
		mLayoutParams.gravity = Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL;
	}
	
	private int getScaledMetric(int value) {
		
		double density = mContext.getResources().getDisplayMetrics().density;
		return (int) Math.ceil(value * density);
	}
	
	private int getUnscaledMetric(int value) {
		
		double density = mContext.getResources().getDisplayMetrics().density;
		return (int) Math.floor(value / density);
	}
}

class MoPubBannerContext extends FREContext implements BannerAdListener {
	
	private MoPubBanner banner;
	
	private MoPubBanner getBanner() {
		
		if (banner == null) {
			
			banner = new MoPubBanner(this.getActivity());
			banner.setBannerAdListener(this);
		}
		return banner;
	}
	
	@Override
	public void dispose() {
		
		if (banner != null) {
			
			banner.setBannerAdListener(null);
			ViewGroup parent = (ViewGroup) banner.getParent();
			
			if (parent != null)
				parent.removeView(banner);
			
			banner.destroy();
		}
	}
	
	// Events
	
	@Override
	public void onBannerLoaded(MoPubView banner) {
		
		dispatchStatusEventAsync("", "bannerLoaded");
	}
	
	@Override
	public void onBannerFailed(MoPubView banner, MoPubErrorCode errorCode) {
		
		dispatchStatusEventAsync("", "bannerFailedToLoad");
	}
	
	@Override
	public void onBannerClicked(MoPubView banner) {
		
		dispatchStatusEventAsync("", "bannerAdClicked");
	}
	
	@Override
	public void onBannerExpanded(MoPubView banner) {
		
		dispatchStatusEventAsync("", "bannerAdExpanded");
	}
	
	@Override
	public void onBannerCollapsed(MoPubView banner) {
		
		dispatchStatusEventAsync("", "bannerAdCollapsed");
	}
	
	// AS interface
	
	private FREFunction banner_init = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				
				MoPubBanner banner = getBanner();
				
				String adUnitId = getStringFromFREObject(args[0]);
				int adSize = args[1].getAsInt();
				
				banner.setAdUnitId(adUnitId);
				banner.setAdSize(adSize);
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getAutorefresh = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				
				return FREObject.newObject(getBanner().getAutorefreshEnabled());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setAutorefresh = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				
				getBanner().setAutorefreshEnabled(getBooleanFromFREObject(args[0]));
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getPositionX = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				return FREObject.newObject(getBanner().getPosX());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getPositionY = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				return FREObject.newObject(getBanner().getPosY());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getFrameWidth = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				return FREObject.newObject(getBanner().getWidth());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getFrameHeight = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				return FREObject.newObject(getBanner().getHeight());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setPositionX = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().setPosX(args[0].getAsInt());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setPositionY = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().setPosY(args[0].getAsInt());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setFrameWidth = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().setFrameWidth(args[0].getAsInt());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setFrameHeight = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().setFrameHeight(args[0].getAsInt());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setAdSize = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().setAdSize(args[0].getAsInt());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getCreativeWidth = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				return FREObject.newObject(getBanner().getAdWidth());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getCreativeHeight = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				return FREObject.newObject(getBanner().getAdHeight());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_loadBanner = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().loadAd();
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_showBanner = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				
				ViewGroup rootContainer = (ViewGroup) ctx.getActivity().findViewById(android.R.id.content);
				rootContainer = (ViewGroup) rootContainer.getChildAt(0);
				rootContainer.addView(getBanner());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_removeBanner = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				ViewGroup parent = (ViewGroup) getBanner().getParent();
				if (parent != null) {
					parent.removeView(getBanner());
				}
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_refresh = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().forceRefresh();
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setKeywords = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().setKeywords(getStringFromFREObject(args[0]));
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_setTesting = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().setTesting(getBooleanFromFREObject(args[0]));
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_getTesting = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				return FREObject.newObject(getBanner().getTesting());
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	private FREFunction banner_moveToDefaultPosition = new BaseFunction() {
		@Override
		public FREObject call(FREContext ctx, FREObject[] args) {
			
			try {
				getBanner().moveToDefaultPosition();
			}
			catch (Exception exception) {
				Log.w("MoPub", exception);
			}
			
			return null;
		}
	};
	
	@Override
	public Map<String, FREFunction> getFunctions() {
		
		Map<String, FREFunction> functionMap = new HashMap<String, FREFunction>();
		functionMap.put("banner_init", banner_init);
		functionMap.put("banner_getAutorefresh", banner_getAutorefresh);
		functionMap.put("banner_setAutorefresh", banner_setAutorefresh);
		functionMap.put("banner_getPositionX", banner_getPositionX);
		functionMap.put("banner_getPositionY", banner_getPositionY);
		functionMap.put("banner_getFrameWidth", banner_getFrameWidth);
		functionMap.put("banner_getFrameHeight", banner_getFrameHeight);
		functionMap.put("banner_setPositionX", banner_setPositionX);
		functionMap.put("banner_setPositionY", banner_setPositionY);
		functionMap.put("banner_setFrameWidth", banner_setFrameWidth);
		functionMap.put("banner_setFrameHeight", banner_setFrameHeight);
		functionMap.put("banner_setAdSize", banner_setAdSize);
		functionMap.put("banner_getCreativeWidth", banner_getCreativeWidth);
		functionMap.put("banner_getCreativeHeight", banner_getCreativeHeight);
		functionMap.put("banner_loadBanner", banner_loadBanner);
		functionMap.put("banner_showBanner", banner_showBanner);
		functionMap.put("banner_removeBanner", banner_removeBanner);
		functionMap.put("banner_refresh", banner_refresh);
		functionMap.put("banner_setKeywords", banner_setKeywords);
		functionMap.put("banner_setTesting", banner_setTesting);
		functionMap.put("banner_getTesting", banner_getTesting);
		functionMap.put("banner_moveToDefaultPosition", banner_moveToDefaultPosition);
		
		return functionMap;
	}
}
